### PR TITLE
Allow multiple files per bucket in hive

### PR DESCRIPTION
**Pull request by losipiuk**

### CODE REVIEW

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java [hunk 1]
```diff
@@ -373,28 +383,88 @@ private void loadPartition(HivePartitionMetadata partition)
...
                 throw new PrestoException(HIVE_INVALID_BUCKET_FILES, format("%s Found sub-directory in bucket directory for partition: %s", CORRUPT_BUCKETING, hiveFileIterator.getPartitionName()));
             }
-            list.add(next);
+            fileStatuses.add(next);
         }
 
-        if (list.size() != bucketCount) {
-            throw new PrestoException(HIVE_INVALID_BUCKET_FILES, format("%s The number of files in the directory (%s) does not match the declared bucket count (%s) for partition: %s", CORRUPT_BUCKETING, list.size(), bucketCount, hiveFileIterator.getPartitionName()));
+        // Sort FileStatus objects (instead of, e.g., fileStatus.getPath().toString). This matches org.apache.hadoop.hive.ql.metadata.Table.getSortedPaths
+        fileStatuses.sort(null);
```
Comment (fiedukow) [inline presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java:399]: Side note: if we want it to be sorted why don't we use `Set<LocatedFileStatus>` in the first place?

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java [hunk 2]
```diff
@@ -373,28 +383,88 @@ private void loadPartition(HivePartitionMetadata partition)
...
+                .collect(Collectors.toList());
+    }
+
+    private List<Set<LocatedFileStatus>> getBucketFilesSetsMulti(List<LocatedFileStatus> sortedFileStatuses)
+    {
+        ImmutableList.Builder<Set<LocatedFileStatus>> list = ImmutableList.builder();
+        ImmutableSet.Builder<LocatedFileStatus> bucketSet = ImmutableSet.builder();
+        String lastBucketId = null;
+
+        for (LocatedFileStatus next : sortedFileStatuses) {
```
Comment (fiedukow) [inline presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java]: Please don't call loop variable `next` when using range for loop.

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java [hunk 3]
```diff
@@ -373,28 +383,88 @@ private void loadPartition(HivePartitionMetadata partition)
...
+        return sortedFileStatuses.stream()
+                .map(ImmutableSet::of)
+                .collect(Collectors.toList());
+    }
+
+    private List<Set<LocatedFileStatus>> getBucketFilesSetsMulti(List<LocatedFileStatus> sortedFileStatuses)
+    {
+        ImmutableList.Builder<Set<LocatedFileStatus>> list = ImmutableList.builder();
+        ImmutableSet.Builder<LocatedFileStatus> bucketSet = ImmutableSet.builder();
+        String lastBucketId = null;
```
Comment (fiedukow) [inline presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java:435]: Optional?

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java [hunk 4]
```diff
@@ -373,28 +383,88 @@ private void loadPartition(HivePartitionMetadata partition)
...
+        // this matches the observed Hive's behaviour during insert. Then files are named using M/R fremwork using
+        // XXXXXXXX_Y_SUFFIX pattern where:
+        // - XXXXXXXX - is task id
+        // - Y - is attempt id
+        // - SUFFIX - is optional copy_N suffix
+
+        // this method will only be used if number of files in directory does not match number of partitions
+        // even if multi file bucketing is enabled.
+
+        String bucketId = Iterables.getFirst(BUCKET_ID_SPLITTER.split(fileName), "");
```
Comment (fiedukow) [inline presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java:463]: I would parse that using regexp and throw an exception if it doesn't match to improve our fail fast capabilities.

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java [hunk 5]
```diff
@@ -373,28 +383,88 @@ private void loadPartition(HivePartitionMetadata partition)
...
-            list.add(next);
+            fileStatuses.add(next);
         }
 
-        if (list.size() != bucketCount) {
-            throw new PrestoException(HIVE_INVALID_BUCKET_FILES, format("%s The number of files in the directory (%s) does not match the declared bucket count (%s) for partition: %s", CORRUPT_BUCKETING, list.size(), bucketCount, hiveFileIterator.getPartitionName()));
+        // Sort FileStatus objects (instead of, e.g., fileStatus.getPath().toString). This matches org.apache.hadoop.hive.ql.metadata.Table.getSortedPaths
+        fileStatuses.sort(null);
+
+        List<Set<LocatedFileStatus>> list = null;
```
Comment (fiedukow) [inline presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java]: Optional?

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java [hunk 6]
```diff
@@ -373,28 +383,88 @@ private void loadPartition(HivePartitionMetadata partition)
...
+        if (fileStatuses.size() == bucketCount) {
+            list = getBucketFilesSetsSingle(fileStatuses);
+        }
+        else if (isMultiFileBucketingEnabled(session)) {
+            list = getBucketFilesSetsMulti(fileStatuses);
+            if (list.size() != bucketCount) {
+                list = null;
```
Comment (fiedukow) [inline presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java]: I don't like idea of using list as control variable.
I would either use well named bool for that or just simply assign value to a list like:

```
if (isMultiFileBucketingEnabled(session) {
  list = getBucketFilesSetsMulti(fileStatuses);
} 
else {
  list = getBucketFilesSetsSingle(fileStatuses);
}
```

and then 

```
if (list.size() != bucketCount) {
  throw new PrestoException(HIVE_INVALID_BUCKET_FILES, "OMG, OMG, OMG, Something went wrong");
}
```

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java [hunk 7]
```diff
@@ -323,48 +330,51 @@ private void loadPartition(HivePartitionMetadata partition)
         // If only one bucket could match: load that one file
```
Comment (fiedukow) [inline presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java:330]: Outdated comment

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientConfig.java [hunk 8]
```diff
@@ -103,7 +103,7 @@ public void testDefaults()
                 .setHdfsPrestoKeytab(null)
                 .setSkipDeletionForAlter(false)
                 .setBucketExecutionEnabled(true)
-                .setBucketWritingEnabled(true));
+                .setMultiFileBucketingEnabled(false));
```
Comment (fiedukow) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientConfig.java]: Maybe we should add some test for `true`?

**Review by fiedukow — APPROVED** [review]
LGTM. Some small optional comments.

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java (hunk 5 above)
Comment (losipiuk) [reply presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java]: done

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java (hunk 2 above)
Comment (losipiuk) [reply presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java]: done

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java (hunk 4 above)
Comment (losipiuk) [reply presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java:463]: naaa

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java (hunk 3 above)
Comment (losipiuk) [reply presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java:435]: This is internal - I will leave null - it is readable in that shape imo.

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java (hunk 6 above)
Comment (losipiuk) [reply presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java]: no - as discussed offline

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java [hunk 9]
```diff
@@ -373,26 +383,86 @@ private void loadPartition(HivePartitionMetadata partition)
...
+        }
+        if (lastBucketId != null) {
+            list.add(bucketSet.build());
+        }
+        return list.build();
+    }
+
+    private String getBucketId(String fileName)
+    {
+        // this matches the observed Hive's behaviour during insert. Then files are named using M/R fremwork using
```
Comment (anusudarsan) [inline presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java:454]: typo. M/R framework

**Pull request closed without merging**